From 04d188efa488431b35e96c5895f18cc2f33b6e44 Mon Sep 17 00:00:00 2001
From: Gatien Chevallier <gatien.chevallier@foss.st.com>
Date: Wed, 15 Feb 2023 11:26:12 +0100
Subject: [PATCH 0248/1695] drivers: stm32_rifsc: add semaphore acquisition

Semaphore on a peripheral should be taken if semaphore mode is enabled,
CID1 is white-listed and resource is secured.

Upstream-Status: Inappropriate [ST downstream https://github.com/STMicroelectronics/optee_os.git ff9306a713a979418cf549ad64153ec9b0c87196 drivers: stm32_rifsc: add semaphore acquisition]
Change-Id: I56a443d1d1e3df00adaa91d9d8d0af811bfd3bdd
Signed-off-by: Gatien Chevallier <gatien.chevallier@foss.st.com>
---
 core/drivers/firewall/stm32_rifsc.c | 30 +++++++++++++++++++++++++++++
 1 file changed, 30 insertions(+)

diff --git a/core/drivers/firewall/stm32_rifsc.c b/core/drivers/firewall/stm32_rifsc.c
index bcb255f8a..5613b82b8 100644
--- a/core/drivers/firewall/stm32_rifsc.c
+++ b/core/drivers/firewall/stm32_rifsc.c
@@ -23,6 +23,7 @@
 #define _RIFSC_RISC_PRIVCFGR0		U(0x30)
 #define _RIFSC_RISC_RCFGLOCKR0		U(0x50)
 #define _RIFSC_RISC_PER0_CIDCFGR	U(0x100)
+#define _RIFSC_RISC_PER0_SEMCR		U(0x104)
 #define _RIFSC_RIMC_CR			U(0xC00)
 #define _RIFSC_RIMC_ATTR0		U(0xC10)
 
@@ -405,6 +406,7 @@ static TEE_Result stm32_risup_cfg(struct rifsc_platdata *pdata,
 	uintptr_t cidcfgr_offset = _OFST_PERX_CIDCFGR * risup->id;
 	uintptr_t offset = sizeof(uint32_t) * (risup->id / _PERIPH_IDS_PER_REG);
 	uint32_t shift = risup->id % _PERIPH_IDS_PER_REG;
+	TEE_Result res = TEE_ERROR_GENERIC;
 
 	if (!risup || risup->id >= drv_data->nb_risup)
 		return TEE_ERROR_BAD_PARAMETERS;
@@ -428,6 +430,34 @@ static TEE_Result stm32_risup_cfg(struct rifsc_platdata *pdata,
 			     BIT(shift));
 	}
 
+	/*
+	 * Take semaphore if the resource is in semaphore mode
+	 * and secured.
+	 */
+	if (SEM_MODE_INCORRECT(risup->cid_attr) ||
+	    !(io_read32(pdata->base + _RIFSC_RISC_SECCFGR0 + offset) &
+	      BIT(shift))) {
+		res = stm32_rif_release_semaphore(pdata->base +
+						  _RIFSC_RISC_PER0_SEMCR +
+						  cidcfgr_offset,
+						  MAX_CID_SUPPORTED);
+		if (res) {
+			EMSG("Couldn't release semaphore for resource %u",
+			     risup->id);
+			return TEE_ERROR_ACCESS_DENIED;
+		}
+	} else {
+		res = stm32_rif_acquire_semaphore(pdata->base +
+						  _RIFSC_RISC_PER0_SEMCR +
+						  cidcfgr_offset,
+						  MAX_CID_SUPPORTED);
+		if (res) {
+			EMSG("Couldn't acquire semaphore for resource %u",
+			     risup->id);
+			return TEE_ERROR_ACCESS_DENIED;
+		}
+	}
+
 	return TEE_SUCCESS;
 }
 
-- 
2.47.2

